### DISCUSSION TITLE
Storing Eloquent relations in Data objects

### DISCUSSION THREAD

**Discussion by iamniels** [spatie/laravel-data#116] [Q&A]
Let's say we have a system which stores orders. Each order is of a brand. Both are represented by Eloquent Models.

We like to create a new order with a request containing the following information:

```php
[
  'brand_name' => 'Varta'
]
```

Without this package we would add a `data()` method to the `FormRequest` class which retrieves the brand: 
```php

public function data()
{
    return new OrderData(
        brand : Brand::whereName($this->input('brand_name'))->sole()
    );
}
```
In the controller we would run for example:

```php
public function create(OrderRequest $request)
{
    return $request->data()->brand->name;
}
```

How can I implement a similar functionality using Data objects?

The only way I could get working was this:

```php
class OrderData extends Data
{
    public function __construct(
        #[WithCast(ModelCast::class, attribute: 'name', input: 'brand_name')]
        public ?Brand $brand,
    )
    {
    }

    public static function rules(): array
    {
        return [
            'brand_name' => 'required|string|max:64'
        ];
    }
}
```
where `ModelCast` retrieves the model.

But there are multiple issues with this code:

- Brand must be defined as nullable, otherwise the request fails on `brand is required`
- The validation rules are spread over both the constructor definition and a separate rules method

So a better approach would be appreciated :-)

**Comment by rubenvanassche** [comment]
Hmm I don't think there's a quick and easy solution for this problem since you're queuing the brand using the `name` field. In such cases I would omit the casts and create a magic creation method like this:

```php
public status function fromRequest(Request $request){
   return new self(
      Brand::whereName($request->input('brand_name'))->sole()
   );
}
```

As for the the `brand` required rule being added, you could add the `WithoutValidation` attribute to skip validation for a data property.

To be honest, in such cases I would manually write a validation rules in a controller or request object and then create the data object in the controller manually. Some cases aren't that simple like this one where we can automatically deduce valuation rules and create data objects without writing additional code.
